app/maptools/measurementmaptool.cpp

```diff
@@ -0,0 +1,274 @@
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "measurementmaptool.h"
+
+MeasurementMapTool::MeasurementMapTool( QObject *parent )
+  : AbstractMapTool{ parent }
+{
+  connect( this, &AbstractMapTool::mapSettingsChanged, this, &MeasurementMapTool::onMapSettingsChanged );
+
+  if ( mapSettings() )
+  {
+    onMapSettingsChanged( mapSettings() );
+  }
+}
+
+MeasurementMapTool::~MeasurementMapTool() = default;
+
+void MeasurementMapTool::addPoint()
+{
+  QgsPoint transformedPoint = mapSettings()->screenToCoordinate( mCrosshairPoint );
+  mPoints.push_back( transformedPoint );
+  rebuildGeometry();
+}
+
+void MeasurementMapTool::removePoint()
+{
+  if ( !mPoints.isEmpty() )
+  {
+    mPoints.pop_back();
+    checkCanCloseShape();
+    rebuildGeometry();
+  }
+}
+
+void MeasurementMapTool::updateDistance()
+{
+  if ( mPoints.isEmpty() )
+  {
+    setLengthWithGuideline( 0.0 );
+    return;
+  }
+
+  checkCanCloseShape();
+
+  QgsPoint lastPoint = mPoints.last();
+  QgsPoint transformedCrosshairPoint = mapSettings()->screenToCoordinate( mCrosshairPoint );
+
+  double calculatedLength = mDistanceArea.measureLength( mRecordedGeometry ) + mDistanceArea.measureLine( transformedCrosshairPoint, lastPoint );
+  setLengthWithGuideline( calculatedLength );
+}
+
+void MeasurementMapTool::checkCanCloseShape()
+{
+  if ( mRecordedGeometry.isEmpty() || mPoints.count() < 3 )
+  {
+    setCanCloseShape( false );
+    return;
+  }
+
+  if ( mPoints.count() >= 3 )
+  {
+    QgsPoint firstPoint = mPoints.first();
+    QPointF firstPointScreen = mapSettings()->coordinateToScreen( firstPoint );
+    double distanceToFirstPoint = std::hypot( mCrosshairPoint.x() - firstPointScreen.x(), mCrosshairPoint.y() - firstPointScreen.y() );
```

Comment:
If you want to have it cleaner, you can create a new function in inpututils, something like `pixelDistanceBetween(QPointF p1, QPointF p2)`